File: solutions/Merge-Two-Sorted-Lists/Merge-Two-Sorted-Lists.py

‘’’
 Definition for singly-linked list.
 class ListNode:
     def __init__(self, x):
         self.val = x
         self.next = None
‘’’
class Solution:
    # @param two ListNodes
    # @return a ListNode
    def mergeTwoLists(self, l1, l2):
        ans=l1
        if l1==None and l2==None:
            return None
        elif l1==None:
            return l2
        elif l2==None:
            return l1
        elif l1.val<l2.val:
            ans=l1
            l1=l1.next
        else:
            ans=l2
            l2=l2.next
        p=ans
        while l1!=None and l2!=None:
            if l1.val<l2.val:
                p.next=l1
                p=p.next
                l1=l1.next
            else:
                p.next=l2
                p=p.next
                l2=l2.next
        if l1!=None:
            p.next=l1
        else:
            p.next=l2
        return ans
            
 

File: solutions/Merge-Two-Sorted-Lists/Merge-Two-Sorted-Lists.cpp

// /**
//  * Definition for singly-linked list.
//  * struct ListNode {
//  *     int val;
//  *     ListNode *next;
//  *     ListNode(int x) : val(x), next(NULL) {}
//  * };
//  */
// class Solution {
// public:
//     ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
//         ListNode *l=NULL;
//         if(!l1 && !l2)
//             return l;
//         else if(!l1)
//             return l2;
//         else if(!l2)
//             return l1;
//         if(l1->val<l2->val)
//         {
//             l=l1;
//             l1=l1->next;
//         }
//         else
//         {
//             l=l2;
//             l2=l2->next;
//         }
//         ListNode *p=l;
//         while(l1 && l2)
//         {
//             if(l1->val<l2->val)
//             {
//                 p->next=l1;
//                 p=l1;
//                 l1=l1->next;
//             }
//             else
//             {
//                 p->next=l2;
//                 p=l2;
//                 l2=l2->next;
//             }
//         }
//         if(l1)
//             p->next=l1;
//         else
//             p->next=l2;
//         return l;
//     }
// };


// 2019-07-15
// O(n)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(!l1)
        {
            return l2;
        }
        else if (!l2)
        {
            return l1;
        }
        ListNode *pre = new ListNode(0);
        ListNode *p = pre;
        while(l1 && l2)
        {
            if (l1->val < l2->val)
            {
                p->next = new ListNode(l1->val);
                l1 = l1->next;
            }
            else
            {
                p->next = new ListNode(l2->val);
                l2 = l2->next;
            }
            p = p->next;
        }
        p->next = l1 ? l1 : l2;
        return pre->next;
    }
};